### ISSUE TITLE
Ingested articles do not load on unstable environment

### ISSUE THREAD

**Issue by GiancarloFusiello** [libero/libero#181]
Demo content has been added to `unstable.libero.pub` (See https://github.com/libero/libero/issues/170 for more details).

The article titles appear on the homepage but do not load individually. HTTP 500 error is displayed.

**Task**
- Identify the cause of the 500 errors and rectify the problem.

**Comment by thewilkybarkid** [comment]
Looks like a bunch of things:

- https://github.com/libero/sample-configuration/pull/50 hasn't been merged so assets aren't moved to their proper location (though original bucket is public, so should work anyway).
- Current references are like https://unstable-jats-ingester-expanded.s3.amazonaws.com/elife-46630-vor-r1/elife-46630-fig1.tif, which doesn't exist?
- Images are TIFF, only JPEG currently supported (so Browser can't currently do anything).

**Comment by GiancarloFusiello** [comment]
>Images are TIFF, only JPEG currently supported (so Browser can't currently do anything).

To further clarify, part of the original spec was to only extract supported document types. The implementation of this can be seen [here](https://github.com/libero/jats-ingester/blob/master/dags/process_elife_zip_dag.py#L24) and below:

```python
ALLOWED_EXTENSIONS = {'.jpg', '.jpeg', '.pdf', '.xml'}
```

@BlueReZZ this is a blocker if we wish to demo articles that reference non supported file types (which is almost all of them).

**Comment by GiancarloFusiello** [comment]
It would appear that this is not actually a bug but we should keep this open until actions to rectify the problem have been agreed and relevant tasks/issues created.

**Comment by BlueReZZ** [comment]
`Content Store` will eventually handle content conversion for display but that might take some time so we can unblock this by transforming TIFF to JPEG with a DAG (or within the existing DAG) similar to how `elife-bot` did it here: https://github.com/elifesciences/elife-bot/blob/develop/activity/activity_ConvertImagesToJPG.py

Perhaps a new task could be raised for adapting that?

**Comment by GiancarloFusiello** [comment]
I've opened a new issue to address this: https://github.com/libero/libero/issues/182

**Comment by GiancarloFusiello** [comment]
@thewilkybarkid Should we allow `.xlsx` and `.docx`  files to be uploaded to the expanded bucket?

**Comment by thewilkybarkid** [comment]
As in assets of those types should be handled? Anything included in the XML needs to be handled.

**Comment by GiancarloFusiello** [comment]
>As in assets of those types should be handled? Anything included in the XML needs to be handled.

Are you suggesting that the jats-ingester should check that each file in the zip archive is referenced in the xml before uploading to the expanded s3 bucket?

**Comment by thewilkybarkid** [comment]
Doesn’t matter whether they go into that bucket or not, but they’ll be ignored as they’re not linked from the XML. (I suppose they could be reference from one of the other files, but due to renaming etc we can’t really support that.)

(Inversely anything referenced in the XML but isn’t present would fail the build.)

**Comment by GiancarloFusiello** [comment]
Ok, I'll create a task to remove the restrictions originally put in place.

**Comment by giorgiosironi** [comment]
New problem: related articles have a wrong `xlink:href` attribute containing just a DOI rather than the absolute URL derived from that DOI:
https://github.com/elifesciences/XML-mapping/blob/master/elife-00666.xml#L501-L510
Short-term solution: strip all these tags in the `jats-ingester`.
Long-term solution: discuss supported formats for related articles. /cc @gnott @libero/elife-production

**Comment by gnott** [comment]
A good observation @giorgiosironi about `<related-article>`. It is stretching the definition of a URI in an example like:

```
<related-article ext-link-type="doi" id="ra1" related-article-type="commentary" xlink:href="10.7554/eLife.18753"/>
```

You could say that a DOI is a unique value, but there's no `scheme` on the data. We do sometimes consider the `ext-link-type="doi"` attribute for what type of link it is (I think usually in `<ext-link>` tags).

In the current XML to JSON conversion process, any DOI that are not URI get a `https://doi.org/` prefix added to the DOI value in order to pass validation.

To evaluate the appropriate value to use in future, I would want to escalate that to @Melissa37 and JATS4R-like people.

**Comment by thewilkybarkid** [comment]
A relative `xlink:href`, if there's no `xml:base`, is relative to the XML file itself.

**Comment by GiancarloFusiello** [comment]
new issue; article `elife-40092-vor-r2` has a reference to a jpeg but the `xlink:href` value does not have a file extension: 
```xml
<graphic xlink:href="elife-40092-resp-fig1" mimetype="image" mime-subtype="jpeg"/>
```
This causes:
```bash
[2019-06-27 10:18:06] request.CRITICAL: Exception GuzzleHttp\Exception\ClientException: "Client error: `GET https://unstable-jats-ingester-expanded.s3.amazonaws.com/elife-40092-vor-r2/elife-40092-resp-fig1` resulted in a `404 Not Found` response: <?xml version="1.0" encoding="UTF-8"?> <Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message> (truncated...) " at /app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php line 113 {"exception":"[object] (GuzzleHttp\\Exception\\ClientException(code: 404): Client error: `GET https://unstable-jats-ingester-expanded.s3.amazonaws.com/elife-40092-vor-r2/elife-40092-resp-fig1` resulted in a `404 Not Found` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message> (truncated...)\n at /app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:113)"} []
```

Should we validated/fix these?

**Comment by Melissa37** [comment]
From the JATS documentation

>`<related-article>` Related Article Information
Description of a journal article related to the content but published separately. May include a link to the related article.

>`xlink:href` (Linking Mechanism)
An address or identifier that permits an XLink application to find a remote resource, for example, a URI (such as a URL or filename) that links to an online site associated with the author or the online definition of a term.
Remarks
As specified in XML Linking Language (XLink) Version 1.0, the value of an `@xlink:href` must be a valid URI.

But the DTD validates with non uri `@xlink:href`

And also, in the documentation non uri `@xlink:href` are shown - probably for an internal link:

```
...
<supplementary-material id="S1" xmlns:xlink="http://www.w3.org/1999/xlink"
xlink:title="local_file" xlink:href="pbio-0020328-t002.xls"
mimetype="application/vnd.ms-excel">
<label>Supporting Material</label>
...
</supplementary-material>
...
```

This matches this linking style we use? I'd be happy to add the full uri with the https etc prefix for DOI related article links, but we cannot really do that for files associated with the main XML because of others hosting the content, such as PMC.

>new issue; article elife-40092-vor-r2 has a reference to a jpeg but the xlink:href value does not have a file extension

In our documentation: https://github.com/elifesciences/XML-mapping/blob/master/elife_file_naming_2016_08_25.md

We ask for a file extension. However, the archive might not have them.
But also the xml will have `.tiff `file extensions but the links on the site are to `.jpeg` so the conversion needs to happen on the XML?

@gnott will know whether he strips them from the XML so they match the website's use (.jpeg are converted from .tiff during hosting and production outputs .tiff)

**Comment by thewilkybarkid** [comment]
> But the DTD validates with non uri `@xlink:href`

The DTD can't validate `@xlink:href`. (And it definitely can't check that it points somewhere real too.)

Though `10.7554/eLife.18753` _is_ a valid URI, it's just relative (and so is made absolute by resolving it against something else, `xml:base` and subsequently the file itself).

**Comment by FAtherden-eLife** [comment]
You can, however, check that the `@xlink:href` points to somewhere real using schematron (although I don't know if XSLT 2 is required) and java.

You can use `java:file-exists()`. I think this requires the full uri path, but this can be derived from the document being validated and the `@xlink:href` in the xml itself. 

To get the full path for the document you can use `base-uri()` (I don't know if this is supported in XSLT 1) and then concatenate (with `@xlink:href`) as necessary to get the full path for the file(s).

Or perhaps we would already know the full path of the document via other means (?), in which case we wouldn't need `base-uri()`.

**Comment by GiancarloFusiello** [comment]
Closing this issue in favour of https://github.com/libero/libero/issues/197
